Lint for invisible Unicode characters other than ZWSP

bugadani opened the pull request:
This PR extends the existing `zero_width_space` lint to look for other invisible characters as well (in this case, `\\u{ad}` soft hyphen.

I feel like this lint is the logical place to add the check, but I also realize the lint name is not particularly flexible, but I also understand that it shouldn't be renamed for compatibility reasons.

Open questions:
 - What other characters should trigger the lint?
 - What should be done with the lint name?
 - How to indicate the change in functionality?

Motivation behind this PR: https://github.com/rust-lang/rust/issues/77417 - I managed to shoot myself in the foot by an invisible character pasted into my test case.

changelog: rename [`zero_width_space`] to [`invisible_characters`] and add SHY and WJ to the list.

rust-highfive commented:
r? @ebroto

(rust_highfive has picked a reviewer for you, use r? to override)

ebroto commented on clippy_lints/src/unicode.rs:
```diff
@@ -91,14 +91,14 @@ fn escape<T: Iterator<Item = char>>(s: T) -> String {
             ZERO_WIDTH_SPACE,
             span,
-            "zero-width space detected",
+            &format!("invisible character detected: {:?}", invisible),
```
As we are going to replace all occurrences, I think we should either specify all the characters that are going to be replaced, or just say `invisible character detected` without specifying which one (they will be visible in the suggestion anyway)

bugadani replied:
Fair enough, thanks!

ebroto reviewed (changes requested):
About the open questions:

>What other characters should trigger the lint?

I'm *really* not an expert in this field. Taking a look at  [this Wikipedia entry](https://en.wikipedia.org/wiki/Whitespace_character#Unicode), there are some other invisible characters (second table) which could be candidates. 
My main concern would be including characters that could be legitimately used in languages I'm not familiar with, so I would prefer to be conservative here. Maybe the `WORD JOINER` one could be a candidate? If we are not sure I would prefer not to include others.

>What should be done with the lint name?

There's a mechanism for renaming lints. You can find an example in #3554. It's a bit old PR, but self-contained, except for the test about renaming lints, you can omit that.

>How to indicate the change in functionality?

A new name for the lint should suffice. What about `invisible_characters` that you used when linting?

bugadani commented:
> I'm _really_ not an expert in this field. Taking a look at [this Wikipedia entry](https://en.wikipedia.org/wiki/Whitespace_character#Unicode), there are some other invisible characters (second table) which could be candidates.

Oh, I intended this question to a broader audience, not placing the burden of figuring out on you. Just in case somebody would have a stronger opinion here.

> My main concern would be including characters that could be legitimately used in languages I'm not familiar with, so I would prefer to be conservative here. Maybe the `WORD JOINER` one could be a candidate? If we are not sure I would prefer not to include others.

I've added WJ, that seems obvious enough. I don't think language support is an issue here - if a character is invisible, I think people should prefer writing out the escape code instead in any language. But I may have wrong assumptions and preferences here.

> There's a mechanism for renaming lints. You can find an example in #3554. It's a bit old PR, but self-contained, except for the test about renaming lints, you can omit that.

Thanks, I've renamed the lint to `invisible_characters`.

ebroto reviewed (approved):
LGTM, let's wait for CI and we can merge it.

>Oh, I intended this question to a broader audience, not placing the burden of figuring out on you

Of course! Sorry if that sounded harsh, it was not the intention.

>I don't think language support is an issue here - if a character is invisible, I think people should prefer writing out the escape code instead in any language

That makes sense, but I would prefer to be cautious. I think adding those two is a good improvement in any case.

ebroto commented:
@bors r+

Thanks!

bors commented:
:pushpin: Commit 572e4c4837e5f955cdc3751b9ad63f0bfb86beac has been approved by `ebroto`

<!-- @bors r=ebroto 572e4c4837e5f955cdc3751b9ad63f0bfb86beac -->
<!-- homu: {"type":"Approved","sha":"572e4c4837e5f955cdc3751b9ad63f0bfb86beac","approver":"ebroto"} -->

bors commented:
:hourglass: Testing commit 572e4c4837e5f955cdc3751b9ad63f0bfb86beac with merge 15a79502397896b0aa53fc44f4f3ae71ccd07e7a...
<!-- homu: {"type":"BuildStarted","head_sha":"572e4c4837e5f955cdc3751b9ad63f0bfb86beac","merge_sha":"15a79502397896b0aa53fc44f4f3ae71ccd07e7a"} -->

bugadani commented:
> Of course! Sorry if that sounded harsh, it was not the intention.

Not at all, no worries. :) Thanks for the review!

bors commented:
:broken_heart: Test failed - [checks-action_test](https://github.com/rust-lang/rust-clippy/runs/1201157673)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang/rust-clippy/runs/1201157673","builder_name":"checks-action_test"} -->

ebroto commented:
@bors retry (missing changelog)

bors commented:
:hourglass: Testing commit 572e4c4837e5f955cdc3751b9ad63f0bfb86beac with merge 9408c68c1da55f8a47cb21e761953db30aea3d39...
<!-- homu: {"type":"BuildStarted","head_sha":"572e4c4837e5f955cdc3751b9ad63f0bfb86beac","merge_sha":"9408c68c1da55f8a47cb21e761953db30aea3d39"} -->

bors commented:
:sunny: Test successful - [checks-action_dev_test](https://github.com/rust-lang/rust-clippy/runs/1201167893), [checks-action_remark_test](https://github.com/rust-lang/rust-clippy/runs/1201166510), [checks-action_test](https://github.com/rust-lang/rust-clippy/runs/1201214299)
Approved by: ebroto
Pushing 9408c68c1da55f8a47cb21e761953db30aea3d39 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"ebroto","base_ref":"master","builders":{"checks-action_test":"https://github.com/rust-lang/rust-clippy/runs/1201214299","checks-action_dev_test":"https://github.com/rust-lang/rust-clippy/runs/1201167893","checks-action_remark_test":"https://github.com/rust-lang/rust-clippy/runs/1201166510"},"merge_sha":"9408c68c1da55f8a47cb21e761953db30aea3d39"} -->
